lib/krane/cluster_resource_discovery.rb

```diff
@@ -18,87 +19,61 @@ def crds
 
     def prunable_resources(namespaced:)
       black_list = %w(Namespace Node ControllerRevision)
-      api_versions = fetch_api_versions
-
-      fetch_resources(namespaced: namespaced).uniq { |r| r['kind'] }.map do |resource|
-        next unless resource['verbs'].one? { |v| v == "delete" }
-        next if black_list.include?(resource['kind'])
-        gvk_string(api_versions, resource)
+      fetch_resources(namespaced: namespaced).map do |resource|
+        next unless resource["verbs"].one? { |v| v == "delete" }
+        next if black_list.include?(resource["kind"])
+        [resource["apigroup"], resource["version"], resource["kind"]].compact.join("/")
       end.compact
     end
 
-    # kubectl api-resources -o wide returns 5 columns
-    # NAME SHORTNAMES APIGROUP NAMESPACED KIND VERBS
-    # SHORTNAMES and APIGROUP may be blank
-    # VERBS is an array
-    # serviceaccounts sa <blank> true ServiceAccount [create delete deletecollection get list patch update watch]
     def fetch_resources(namespaced: false)
-      command = %w(api-resources)
-      command << "--namespaced=#{namespaced}"
-      raw, err, st = kubectl.run(*command, output: "wide", attempts: 5,
-        use_namespace: false)
-      if st.success?
-        rows = raw.split("\n")
-        header = rows[0]
-        resources = rows[1..-1]
-        full_width_field_names = header.downcase.scan(/[a-z]+[\W]*/)
-        cursor = 0
-        fields = full_width_field_names.each_with_object({}) do |name, hash|
-          start = cursor
-          cursor = start + name.length
-          # Last field should consume the remainder of the line
-          cursor = 0 if full_width_field_names.last == name.strip
-          hash[name.strip] = [start, cursor - 1]
-        end
-        resources.map do |resource|
-          resource = fields.map { |k, (s, e)| [k.strip, resource[s..e].strip] }.to_h
-          # Manually parse verbs: "[get list]" into %w(get list)
-          resource["verbs"] = resource["verbs"][1..-2].split
-          resource
-        end
-      else
-        raise FatalKubeAPIError, "Error retrieving api-resources: #{err}"
+      responses = {}
+      Krane::Concurrency.split_across_threads(api_paths) do |path|
+        responses[path] = fetch_api_path(path)["resources"] || []
       end
```

Comment:
The only concrete reading I found regarding this is https://ivoanjo.me/blog/2019/08/26/quick-tip-unsafe-concurrent-ruby-hash-access/
While the above bug doesn't to happen in MRI (due to GIL), it's still making me nervous about having multiple threads assign to a single hash.
I wonder if it isn't worth the effort of implementing `collect_across_threads` (rough example below).

```ruby
    def self.collect_across_threads(all_work, max_threads: MAX_THREADS, &block)
      return [] if all_work.empty?
      raise ArgumentError, "Block of work is required" unless block_given?

      slice_size = ((all_work.length + max_threads - 1) / max_threads)
      threads = []
      all_work.each_slice(slice_size).each_with_index do |slice, index|
        threads << Thread.new do
          slice.map { |s| yield s }
        end
      end
      collection = []
      threads.each do |t|
        collection << t.value
      end
      collection
    end
```

Semantically, it seems reasonable to differentiate between `split_across_threads` and `collect_across_threads` since they imply per-object side effects in the former and aggregation of data in the latter.

Reply:
I did some digging. It sounds like the code I've got is fine in MRI https://bugs.ruby-lang.org/issues/14735#note-1 If that makes you nervious, we could always use the thread-safe hash in `concurrent-ruby` which we pull in because we grab activesupport...

Reply:
I'm leaning towards concurrent-ruby. As your linked article suggests, the fact that Hash is thread-safe is dependent on using MRI and using specific classes for the hash key. It's also suggested that these are implementation details and we shouldn't rest our assumptions on them remaining constant. 

However, if we observe too large a time penalty spent synchronizing we can always take a look again